Simplify NAT46x64,recorder tests

joestringer opened the pull request:
Requires https://github.com/cilium/cilium/pull/32070 to test.

Review commit-by-commit to reason through the changes over the course of the PR.

Previously, these tests would emit very verbose output which was hard to
interpret when it failed, for instance a test may fail and this is the output:

```
++[01:02:57] docker exec -t lb-node ip -o -4 a s eth0
++[01:02:57] awk '{print $4}'
++[01:02:57] head -n1
++[01:02:57] cut -d/ -f1
+[01:02:57] LB_NODE_IP=172.12.42.2
+[01:02:57] ip r a 10.0.0.8/32 via 172.12.42.2
++[01:02:57] seq 1 10
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null '[fd00:cafe::1]:80'
++[01:02:57] seq 1 10
+[01:02:57] for i in $(seq 1 10)
+[01:02:57] curl -s -o /dev/null 10.0.0.8:80
+[01:02:57] echo 'Failed 1'
Failed 1
+[01:02:57] exit -1
Error: Process completed with exit code 255.
```

If you understand what the test is trying to do and where in the test that
these commands were run, then you might be able to debug it. But for most
mortals, this output was not very easy to understand.

With this PR, the tests are reworked to only print verbose output when the
environment variable `V=1` is set, and by default it prints a much terser
set of output, such as:

```
[./test/nat46x64/test.sh:228]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
9c536a08a4bdb8751cdc2337503c5864e1fd1984619635b466cf4352d99d71b2
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:229]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Maglev
[./test/nat46x64/test.sh:233]     Installing Cilium with Mode:TC  Algorithm:Random        Recorder:Disabled
cilium-lb
894298c1821b2b536f6744c78aa886bc8f3ab7f4a291b82d9ccb6677d253f26b
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:234]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Random
[./test/nat46x64/test.sh:238]     Configuring service 10.0.0.8:80 -> 172.12.42.3:80
Deprecation warning: --id parameter will change from int to string in v1.14
Creating new service with id '2'
Added service with 1 backends
[./test/nat46x64/test.sh:244]     Installing route: 'ip -4 r a 10.0.0.8/32 via 172.12.42.2'
[./test/nat46x64/test.sh:246]     Checking connectivity via [fd00:cafe::1]:80 -> 172.12.42.3:80
[./test/nat46x64/test.sh:249]     Checking connectivity via 10.0.0.8:80 -> 172.12.42.3:80
[./test/nat46x64/test.sh:252]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
89f2ad155a0a5ae0922ba741baa8d6186d4f87ee4a3e019d5845a1847b6f5f1d
cilium: daemon unreachable
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:253]     Testing service [fd00:cafe::1]:80 -> 172.12.42.3:80 via TC + Maglev
Deprecation warning: --id parameter will change from int to string in v1.14
Service 1 deleted successfully
Deprecation warning: --id parameter will change from int to string in v1.14
Service 2 deleted successfully
```

Or, on failure:

```
$ sudo ./test/nat46x64/test.sh
[./test/nat46x64/test.sh:197]     Installing Cilium with Mode:TC  Algorithm:Maglev        Recorder:Disabled
cilium-lb
ccb3d8f50c6d15a016035eb81c12e62efa04ef79cad62a00a1c1b9bd529ca58c
cilium: daemon unreachable
OK
[./test/nat46x64/test.sh:199]     Configuring service 10.0.0.4:80 -> [2001:db8:1::3]:80
Deprecation warning: --id parameter will change from int to string in v1.14
Creating new service with id '1'
Added service with 1 backends
[./test/nat46x64/test.sh:209]     Testing service 10.0.0.4:80 -> [2001:db8:1::3]:80 via TC + Maglev
[./test/nat46x64/test.sh:210]     Installing route: 'ip -4 r a 10.0.0.4/32 via 172.12.42.2'
[./test/nat46x64/test.sh:211]     Failed connection from localhost to 10.0.0.4:80 (attempt 5/10)
```

The annotated lines in the square brackets also provide a test file and line number where the failure occurred to make it easier to track down the failure and attempt to reproduce.

joestringer committed:
nat46x64: Fix shellcheck warnings

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Refactor maglev map sanity checking

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Refactor docker env init

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Clean up environment by default

Avoid leaving state left behind unless explicitly requested.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Refactor common configurations

Apply all of the common configurations with a variable from the top of
the file, to simplify tracking down the different configurations
throughout the test.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Refactor waiting for service to be ready

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Add IPv4 service readiness check

This mirrors the exact same pattern in the IPv6->IPv6 case, see commit
129bc1e3fc78 ("test,l4lb: Change expected error in L4LB test") for more
details.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Always enable nat46x64 gateway

For some reason, the installations of Cilium in this test seem to just
randomly decide whether they're going to install with nat46x64 gateway
enabled or not. I can't tell if this is intentional, so this commit
changes all installs in this file to enable the feature.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Centralize configuration, improve output

Make the test output more clear by echoing to the terminal a message
whenever reconfiguring Cilium.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Remove redundant compile tests

These configurations were already being run elsewhere in the test.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Add debugging steps and better tracing

Improve the tracing and failure output in the test.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Refactor local node route configuration

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Quieten test output by default

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Print cilium-agent logs on install fail

If cilium-agent doesn't come up ready, print the logs to the terminal
and fail out. This should make a common class of possible failure
conditions much easier to debug.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Rework variables for primary test

The initial test performs NAT46, Service 6to6, then does some validation
around the scenarios. The second commit does the inverse, NAT64 and
4to4 with the same basic checks afterwards.

This commit reworks the initial set of tests so that the code can be
shared for both tests in a subsequent commit.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Move NAT46 test into function

No functional changes. This will be reused in the next commit by the
NAT64 test.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Combine NAT46 and NAT64 tests

These tests were running practically the same logic, just duplicated.
Now that the previous commit abstracted away the differences between the
frontends and backends in the tests into common variables, we can
convert the core of the test into a function and call it with different
variables to run the different tests.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Improve connectivity check fail reports

Centralize the connectivity checks into a common function and update the
output message to fail more clearly with a line number and other details
about how the connection is being generated.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Split data/test logic for recorder

Move the declaration of the arrays of filters above the test code to
make it easier to read through what the test is doing. While we're at
it, silence unnecessary test output unless in verbose mode (env V=1).

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Test recorder filter installation

Add a couple of checks to ensure that all of the requested filters end
up in the BPF maps.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Only print line numbers

BASH_LINENO is an array, and only the first index is useful to
understand the line number, so only format that part.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
nat46x64: Hold environment by default

Apparently the CI relies on this script only cleaning up after itself
upon success, so switch the default behaviour back to this. On a local
TTY, the developer will be asked whether to clean up the environment if
there's a failure. Otherwise it'll only be cleaned up on success.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on test/nat46x64/test.sh:
```diff
@@ -1,423 +1,411 @@
 #!/usr/bin/env bash
 
 PS4='+[\t] '
```
So heads up by default these timestamps won't show up when you run the script, unless you set the env var `V=1` like `V=1 sudo -E ./test/nat46x64/test.sh`. I think this is fine as a tradeoff, in GitHub actions we have access to timestamps anyway through the UI:

![image](https://github.com/cilium/cilium/assets/1243336/316d335a-dbed-48a8-9798-e280eccbb73d)

tklauser commented on test/nat46x64/test.sh:
```diff
@@ -1,423 +1,411 @@
+    echo "Found filters:"
+    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+    fatal "Recorder filters did not match expected list"
+fi
```
nit: these two blocks look similar enough that they might be refactored into a function as well.

joestringer replied:
I agree. I did try this earlier when developing these tests, but being able to properly pass the array contents into a function was a little tricky. I can try to take another stab at this, but if it's not trivial then I'll probably skip this feedback. IMO we're reaching diminishing returns on iterating on these steps at this point.

Recorder is also a bit weird to be combined into the nat46x64 test, it seems to be entirely unrelated; I wouldn't mind if it were in its own file. But a bit more refactoring is necessary to get that done. We can continue to iterate in the tree.

joestringer replied:
Looks easy enough. I'll add an extra commit on top:

```
diff --git a/test/nat46x64/test.sh b/test/nat46x64/test.sh
index 6366f448a838..75a5aa0af4e0 100755
--- a/test/nat46x64/test.sh
+++ b/test/nat46x64/test.sh
@@ -258,6 +258,23 @@ function test_services {
     ${CILIUM_EXEC} cilium-dbg service delete 2
 }

+# Check whether the list of filters matches the expected list.
+#
+# $1 - ID
+# $@ - Filters
+function check_recorder_list {
+    ID=$1
+    shift
+    num_filters=$#
+    if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:${ID}" | wc -l) -ne "$num_filters" ]; then
+        echo "Expected filters:"
+        echo "$@" | nl -bn
+        echo "Found filters:"
+        ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
+        fatal "Recorder filters did not match expected list"
+    fi
+}
+
 force_cleanup 2>&1 >/dev/null
 initialize_docker_env
 trap cleanup EXIT
@@ -383,23 +400,8 @@ ${CILIUM_EXEC} \
     cilium-dbg recorder update --id 2 --caplen 100 \
         --filters="$(printf '%s,' "${RECORDER_FILTERS_IPV6[@]}" | sed 's/,*$//')"

-info "Checking the list of filters in IPv4 recorder ID:1"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:1" | wc -l) -ne ${#RECORDER_FILTERS_IPV4[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV4[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
-
-info "Checking the list of filters in IPv6 recorder ID:2"
-if [ $(${CILIUM_EXEC} cilium-dbg bpf recorder list | grep "ID:2" | wc -l) -ne ${#RECORDER_FILTERS_IPV6[@]} ]; then
-    echo "Expected filters:"
-    echo "${RECORDER_FILTERS_IPV6[@]}" | nl -bn
-    echo "Found filters:"
-    ${CILIUM_EXEC} cilium-dbg bpf recorder list | nl -bn
-    fatal "Recorder filters did not match expected list"
-fi
+check_recorder_list 1 "${RECORDER_FILTERS_IPV4[@]}"
+check_recorder_list 2 "${RECORDER_FILTERS_IPV6[@]}"

 trace_exec ${CILIUM_EXEC} cilium-dbg recorder list
 trace_exec ${CILIUM_EXEC} cilium-dbg bpf recorder list
```

tklauser reviewed (approved):
Thanks for splitting this up into nicely documented commits (took a while though to read through all of them :wink:). One non-blocking nit about a potential refactoring inline.

aspsk commented on test/nat46x64/test.sh:
```diff
@@ -53,12 +53,11 @@ cilium_install \
 NGINX_PID=$(docker inspect nginx -f '{{ .State.Pid }}')
-WORKER_IP4=$(nsenter -t $NGINX_PID -n ip -o -4 a s eth0 | awk '{print $4}' | cut -d/ -f1 | head -n1)
-WORKER_IP6=$(nsenter -t $NGINX_PID -n ip -o -6 a s eth0 | awk '{print $4}' | cut -d/ -f1 | head -n1)
-WORKER_MAC=$(nsenter -t $NGINX_PID -n ip -o l show dev eth0 | grep -oP '(?<=link/ether )[^ ]+')
```
Had this line caused a warning? Or just an unquoted variable?

joestringer replied:
I think `WORKER_MAC` was entirely unused. Otherwise the `WORKER_IP*` vars had an unquoted `$NGINX_PID` yeah.

joestringer replied:
If your question was in reference to a particular commit, I could take a deeper look at that commit and the context to make sure that individual change didn't miss something.

aspsk commented on test/nat46x64/test.sh:
```diff
@@ -60,7 +60,30 @@ function initialize_docker_env {
 }
 
+function force_cleanup {
+    ${CILIUM_EXEC} cilium-dbg service delete 1 || true
```
I usually use `smth || :` instead :)

joestringer replied:
TIL. Tidier but harder to read if you're less familiar with bash, which applies to most contributors in the codebase.

aspsk commented on test/nat46x64/test.sh:
```diff
@@ -80,6 +80,17 @@ function cleanup {
+    # Try and sleep until the LB2 comes up, seems to be no other way to detect when the service is ready.
+    set +e
+    for i in $(seq 1 10); do
+        curl -s -o /dev/null "${LB_ALT}" && break
```
the `URL` is passed via a function parameter, however, "${LB_ALT}" (without :80) is hardcoded here. Should be `$1` instead or
```
local url="$1"
...
curl ... $url
```

joestringer replied:
Ah, good find, let me fix that up.

aspsk reviewed (commented):
LGTM, see a few minor comments

aspsk reviewed (changes requested):
Hereby I am marking my previous review as "request for changes"

joestringer committed:
nat46x64: Refactor recorder list checking

Move the tests for the recorder lists into a dedicated function.

Suggested-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>
